torch/csrc/jit/tracer.cpp

```diff
@@ -378,8 +378,18 @@ void addInputs(
 
 void addInputs(Node* n, const char* name, at::TensorList value) {
   Graph* g = n->owningGraph();
-  Node* list_node = g->insertNode(
-      g->createList(DynamicType::get(), fmap(value, getValueTrace)));
+  TypePtr typ = DynamicType::get();
+  // NB: for TensorList, we create the node depending on the values inside,
```

Comment:
This will not work correctly because List[T] <: List[Optional[T]] is false. addInputs is only called from generated code where it is known whether the operator takes a Tensor[] or Tensor?[]. This addInputs can have a `bool allow_undefined` flag which will always produce an Optional type.

Reply:
I think addInputs is called from VariableType_x.cpp and inside it for the index op, it only takes input of a TensorList, so in here I chose to iterate the TensorList and see if there's undefined tensor inside, if there is, createList(Optional[Tensor]), otherwise createList(DynamicType::get()), it does not use the subtyping of List so I think should be fine.  Not sure how can we know the operator takes a `Tensor?[]`?

Reply:
This does not work. If an ops's schema says that it takes a `Tensor?[]` then addInputs _must_ provide a list of that type, _regardless of what the actual members of the list are_. However, all calls to addInputs are always generated by gen_jit_dispatch, where _it is known what the type of this list is_. So it is possible to inform this addInputs about what the type of the list should be from the caller.

Reply:
I see, I think what you mean is the gen_variable_type.py, I will make changes to this AddInputs to allow the `bool allow_undefined` flag to be passed in to handle the case where we have `Tensor?[]` to create a list of that type.